### ISSUE TITLE
Browser.Types.Worker incompatibility

### ISSUE THREAD

**Issue by laurentpayot** [fable-compiler/fable-browser#137]
Hi, I’m trying to use a web worker of *module* type in an Elmish app: 

```fsharp
open System
open Fable.Core.JsInterop
open Elmish
open Browser.Worker
open Browser.Types

// ...

type ModuleWorkerOptions() =
    interface WorkerOptions with
        member _.``type`` = WorkerType.Module

let worker = Worker.Create("worker.js", new ModuleWorkerOptions())

let onWorkerEvent
    (worker: Worker)
    (eventName: string)
    (toMsg: string -> 'msg)
    : ('msg Dispatch -> IDisposable) =
    let start dispatch =
        // TODO
        worker.onmessage <-
            (fun event ->
                let str = event.data?value
                dispatch (toMsg str))

        { new IDisposable with
            member _.Dispose() = ()
        }

    start

let subscriptions (model: Model) : Msg Sub = [
    ([ "AppUser" ], onWorkerEvent worker "AppUser" GotJsonAppUser)
]

//...

```

but I’m getting the following error:

```txt
The type 'Browser.Types.Worker (Browser.Worker, Version=1.3.0.0, Culture=neutral, PublicKeyToken=null)' is not compatible with the type 'Browser.Types.Worker (Browser.Dom, Version=2.16.0.0, Culture=neutral, PublicKeyToken=null)
```

I could get a non-module, *classic* web worker to work with `onWorkerEvent` with
```fsharp
let worker = Worker.Create("worker.js") :?> Worker
```
but I really need a module web worker.

I’m also wondering:
- Is defining `worker.onmessage` with a back arrow `<-` the right way to do it? That’s the only way I could make it work but it seems a bit abrupt to me.
- Is there a more direct way to set a module options instead of defining a new `ModuleWorkerOptions` interface? 

So basically I’m craving for web worker examples :pray:

**Comment by MangelMaxime** [comment]
> * Is defining `worker.onmessage` with a back arrow `<-` the right way to do it? That’s the only way I could make it work but it seems a bit abrupt to me.

This is the correct way, `onmessage` is property that you need to set.

```js
monWorker.onmessage = function (e) {
  resultat.textContent = e.data;
  console.log("Message reçu depuis le worker");
};
```

> * Is there a more direct way to set a module options instead of defining a new `ModuleWorkerOptions` interface?

You can use `jsOptions` or `createEmpty` for this kind of definition. Learn more about it on [Fable docs](https://fable.io/docs/javascript/features.html#plain-old-javascript-objects)

Between, these 2 options I personally prefer `jsOptions`

Hum, perhaps we need to release a new version of `Browser.Types.Worker` to resolve some incompatibility ? 🤔

**Comment by MangelMaxime** [comment]
Looking into the repository source it seems like Worker API is defined at 2 places:

- Fable.Browser.Worker
- Fable.Browser.Dom

You are probably trying to use an API defined in Fable.Browser.Dom by passing it types coming from Fable.Browser.Worker? Does `dotnet build` works? If the error reported only on compilation or also in the IDE?

**Comment by MangelMaxime** [comment]
I believe we should have the Worker API in a single place, but we would need to check if there are duplicates etc.

To respect the philosophy of this repository it should probably all be in the Worker project.

**Comment by laurentpayot** [comment]
Thanks @MangelMaxime for your answer.

> You are probably trying to use an API defined in Fable.Browser.Dom by passing it types coming from Fable.Browser.Worker?

Yes I do, because I could not find a `Worker` type for my `onWorkerEvent` function signature anywhere but in `Browser.Types`. Is there a way to use a type from `Fable.Browser.Worker`?

> Does dotnet build works? If the error reported only on compilation or also in the IDE?

The error is reported both at compilation time and at writing time in the IDE (VS Code with Ionide plugin).

**Comment by laurentpayot** [comment]
> You can use jsOptions or createEmpty for this kind of definition

Thanks a lot, I ended up using the following:
```fsharp
let options = jsOptions<WorkerOptions> (fun options -> options.``type`` <- Module)
let worker = Worker.Create("worker.js", options) :?> Worker
```

Note that without the `:?> Worker` part the error is still here, because of the duplicated definitions I guess…

**Comment by MangelMaxime** [comment]
I removed the Worker code from Fable.Browser.Dom.

You should not have conflicts anymore now.

**Comment by laurentpayot** [comment]
I can confirm `:?> Worker` is not needed anymore. Awesome. Thanks Maxime!
